### PR TITLE
Full table replication

### PR DESCRIPTION
**Pull request by mdelaurentis**
The main motivation is to allow full table replication by adding version numbers to records and issuing ACTIVATE_VERSION messages, but I made several other changes here also.

1. Refactor State and StreamState so that their constructors don't do anything other than set properties. Now use `State.from_dict()` to build the state from the raw dictionary. I did this because when I went to add a property to StreamState, I found the constructor to be confusing since it was doing so much stuff.
2. Use tap_stream_id to identify the stream in the state.
3. Use tap_stream_id to indicate the current stream in the state.
4. Add a "version" property to StreamState.
5. Add the version number to each record.
6. Emit an ActivateVersion message for each table. For incremental tables, it comes at the beginning of the sync. For full-table replication, it comes at the end.
7. Added unit tests for each of four cases:
    1. Incremental syncing with no initial state
    1. Incremental syncing with an initial state
    1. Full-table syncing with no initial state
    1. Full-table syncing with an initial state
8. In discovery mode:
    1. Mark all nodes with `"selected": false`
    2. Set "schema" to be equal to the table name

### CODE REVIEW

File: tap_mysql/__init__.py [hunk 1]
```diff
@@ -82,13 +82,48 @@ class InputException(Exception):
...
+    '''Represents the state for a single stream.
+
+    The state for a stream consists of four properties:
+
+      * tap_stream_id - the identifier for the stream
+      * replication_key (optional, string) - name of the field used for
+        bookmarking
+      * replication_key_value (optional, string or int) - current value of
+        the bookmark
+      * version (optional, int) - the version number of the stream.
```
Comment (bi1yeu) [inline tap_mysql/__init__.py]: @mdelaurentis this is marked as `optional` -- in what cases would a stream state exist without a version?

File: tap_mysql/__init__.py [hunk 2]
```diff
@@ -82,13 +82,48 @@ class InputException(Exception):
...
+        '''
+        result = StreamState(catalog_entry.tap_stream_id)
+        if catalog_entry.replication_key:
+            result.replication_key = catalog_entry.replication_key
+            if raw and raw.get('replication_key') == catalog_entry.replication_key:
+                result.replication_key_value = raw['replication_key_value']
+        if raw and 'version' in raw:
+            result.version = raw['version']
+        else:
+            result.version = int(time.time()) * 1000
```
Comment (bi1yeu) [inline tap_mysql/__init__.py]: first mult by 1000 then cast to int to preserve millisecond resolution

File: tap_mysql/__init__.py [hunk 3]
```diff
@@ -390,13 +438,12 @@ def sync_table(connection, db, table, columns, state):
...
         params = {}
-        stream_state = state.get_stream_state(table)
-        if stream_state and stream_state.replication_key_value is not None:
+        if stream_state.replication_key_value is not None:
             key = stream_state.replication_key
             value = stream_state.replication_key_value
             select += ' WHERE `{}` >= %(replication_key_value)s ORDER BY `{}` ASC'.format(key, key)
             params['replication_key_value'] = value
-        elif stream_state:
+        elif  stream_state.replication_key is not None:
```
Comment (bi1yeu) [inline tap_mysql/__init__.py]: extra space

File: tap_mysql/__init__.py (hunk 1 above)
Comment (mdelaurentis) [inline tap_mysql/__init__.py]: I guess it wouldn't. I'll remove the optional.

File: tap_mysql/__init__.py (hunk 2 above)
Comment (mdelaurentis) [inline tap_mysql/__init__.py]: True. Thanks.

File: tap_mysql/__init__.py (hunk 3 above)
Comment (mdelaurentis) [inline tap_mysql/__init__.py]: k

**Pull request merged**